tutor/src/screens/gradebook/weights-ux.js

```diff
@@ -0,0 +1,203 @@
+import { observable, computed, action } from 'mobx';
+import { sum, toArray, flow, inRange, keys, isEqual, pick, invert, mapValues, isNaN, partial } from 'lodash';
+
+const CELL_AVERAGES_SINGLE_WIDTH = 80;
+
+const DEFAULTS = {
+  homework_scores: 100,
+  homework_progress: 0,
+  reading_scores: 0,
+  reading_progress: 0,
+};
+
+const RECOMMENDED = {
+  homework_scores: 50,
+  homework_progress: 0,
+  reading_scores: 0,
+  reading_progress: 50,
+};
+
+const CW = {
+  homework_score_weight: 'homework_scores',
+  homework_progress_weight: 'homework_progress',
+  reading_score_weight: 'reading_scores',
+  reading_progress_weight: 'reading_progress',
+};
+
+const WC = invert(CW);
+
+const PERCENT_KEYS = keys(DEFAULTS);
+const WEIGHT_KEYS = keys(CW);
+
+const MIN = 0;
+const MAX = 100;
+const RANGE = MAX - MIN;
+
+const weightToPercent = (weight) => ((weight || MIN) * RANGE );
+const percentToWeight = (percent) => ((percent || MIN) / RANGE );
+
+const weightsToPercents = (weights) => mapValues(WC, (c) => weightToPercent(weights[c]));
+const percentsToWeights = (percents) => mapValues(CW, (w) => percentToWeight(percents[w]));
+
+const stringToInt = (string) => {
+  const int = parseInt(string);
+
+  if (isNaN(int)) {
+    return 0;
+  }
+
+  return int;
+};
+
+export default class ScoresReportWeightsUX {
+
+  PERCENT_KEYS = PERCENT_KEYS;
+  WEIGHT_KEYS = WEIGHT_KEYS;
+
+  @observable homework_scores;
+  @observable homework_progress;
+  @observable reading_scores;
+  @observable reading_progress;
+
+  @observable hasTouched = false;
+  @observable showWeightsModal = false;
+
+  constructor(scoresUx) {
+    this.scoresUx = scoresUx;
+  }
+
+  @computed get course() {
+    return this.scoresUx.course;
+  }
+
+  @computed get isBusy() {
+    return Boolean(this.course.api.isPending || this.course.scores.api.isPending);
+  }
+
+  @computed get savingButtonText() {
+    if (this.course.scores.api.isPending) {
+      return 'Recalculating scores…';
+    } else {
+      return 'Saving…';
+    }
+  }
+
+  /**
+   * Show Weights modal
+   */
+  @action showWeights() {
+    this.showWeightsModal = true;
+    Object.assign(this, this.currentPercents);
+  }
+
+  /**
+   * Hide Weights modal
+   */
+  @action hideWeights() {
+    this.showWeightsModal = false;
+  }
+
+  @action.bound onSaveWeights() {
+    const { course, currentPercents } = this;
+    Object.assign(course, this.nextWeights);
+    course
+      .save()
+      .then(() => course.scores.fetch())
+      .catch(() => {
+        // reset course weights to previous weight values
+        Object.assign(course, percentsToWeights(currentPercents));
+      })
+      .then(() => {
+        this.isSetting = false;
+      });
+  }
+
```

Comment:
this is fine, but we also support async await, it could be:
```
  @action.bound async onSaveWeights() {
     const { course, currentPercents } = this;
     Object.assign(course, this.nextWeights);
     try {
       await course.save()
       await course.scores.fetch())
     } catch () {
         // reset course weights to previous weight values
         Object.assign(course, percentsToWeights(currentPercents));
      }
      finally {
        this.isSetting = false;
     }
   }

```

Reply:
:+1: